pkg/systemstatsmonitor/osfeature_collector.go

```diff
@@ -0,0 +1,108 @@
+package systemstatsmonitor
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+	ssmtypes "k8s.io/node-problem-detector/pkg/systemstatsmonitor/types"
+	"k8s.io/node-problem-detector/pkg/util"
+	"k8s.io/node-problem-detector/pkg/util/metrics"
+	"k8s.io/node-problem-detector/pkg/util/metrics/system"
+)
+
+type osFeatureCollector struct {
+	osFeature *metrics.Int64Metric
+}
+
+func NewOsFeatureCollectorOrDie(osFeatureConfig *ssmtypes.OsFeatureStatsConfig) *osFeatureCollector {
+	oc := osFeatureCollector{nil}
+	var err error
+	// Use metrics.Sum aggregation method to ensure the metric is a counter/cumulative metric.
+	if osFeatureConfig.MetricConfigs["system/os_feature"].DisplayName != "" {
+		oc.osFeature, err = metrics.NewInt64Metric(
+			metrics.OsFeatureID,
+			osFeatureConfig.MetricConfigs[string(metrics.OsFeatureID)].DisplayName,
+			"Guest OS Features",
+			"1",
+			metrics.LastValue,
+			[]string{featureLabel, "value"})
+		if err != nil {
+			glog.Fatalf("Error initializing metric for system/os_feature: %v", err)
+		}
+	}
+	return &oc
+}
+
+func (os *osFeatureCollector) recordFeaturesFromCmdline(cmdlineArgs []system.CmdlineArg) {
+	var featuresMap = map[string]int64{
+		"KTD":                    0,
+		"UnifiedCgroupHierarchy": 0,
+		"ModuleSigned":           0,
+		"LoadPinEnabled":         0,
+	}
+	for _, cmdlineArg := range cmdlineArgs {
+		if cmdlineArg.Key == "csm.enabled" {
+			featuresMap["KTD"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+		if cmdlineArg.Key == "systemd.unified_cgroup_hierarchy" {
+			featuresMap["UnifiedCgroupHierarchy"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+		if cmdlineArg.Key == "module.sig_enforce" {
+			featuresMap["ModuleSigned"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+		if cmdlineArg.Key == "loadpin.enabled" {
+			featuresMap["LoadPinEnabled"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+	}
+	// Record the values
+	os.osFeature.Record(map[string]string{featureLabel: "KTD"}, featuresMap["KTD"])
+	os.osFeature.Record(map[string]string{featureLabel: "UnifiedCgroupHierarchy"}, featuresMap["UnifiedCgroupHierarchy"])
+	if featuresMap["ModuleSigned"] == 1 && featuresMap["LoadPinEnabled"] == 1 {
+		os.osFeature.Record(map[string]string{featureLabel: "KernelModuleIntegrity"}, 1)
+	} else {
+		os.osFeature.Record(map[string]string{featureLabel: "KernelModuleIntegrity"}, 0)
+	}
+}
+
+func (os *osFeatureCollector) recordFeaturesFromModules(modules []system.Module) {
+	defaultModules, err := util.GetDefaultModules()
+	if err != nil {
+		glog.Warningf("Error retrieving default modules %v", err)
+	}
+	unknownModules := []string{}
+	for _, module := range modules {
+		if module.OutOfTree || module.Proprietary || module.Unsigned {
+			if !find(module.ModuleName, defaultModules) {
+				unknownModules = append(unknownModules, module.ModuleName)
+			}
+		}
+	}
+	if len(unknownModules) > 0 {
```

Comment:
I am actually not sure that UnknownModules and osFeatures are saying the same thing, maybe consider separate it to an individual metric and can also add module's status into the metric

Reply:
it would be more like if the os has UnknownModules installed.